oolab/src/main/java/agh/ics/oop/RectangularMap.java

```diff
@@ -3,55 +3,32 @@
 import java.util.ArrayList;
 import java.util.List;
 
-public class RectangularMap implements IWorldMap{
+public class RectangularMap extends AbstractWorldMap{
 
-    private List<Animal> animals = new ArrayList<>();
     private static final Vector2d LOWER_BOUND = new Vector2d(0,0);
-    private static Vector2d UPPER_BOUND; /* Yup, I know it's not a constant value,
-    but I will still use it like that, because it's convenient */
+    private static Vector2d upper_bound; /* upper bound inclusive */
 
-    RectangularMap(int width, int height){
-        UPPER_BOUND = new Vector2d(width-1, height-1);
+    public RectangularMap(int width, int height){
+        upper_bound = new Vector2d(width-1, height-1);
     }
 
     @Override
     public boolean canMoveTo(Vector2d dest){
-        if(!(LOWER_BOUND.precedes(dest))) return false;
-        if(!(UPPER_BOUND.follows(dest))) return false;
-        if(isOccupied(dest)) return false;
-        return objectAt(dest) == null;
-    }
-
-    @Override
-    public boolean place(Animal animal){
-        if(!canMoveTo(animal.getPosition()))
-            return false;
-        animals.add(animal);
-        return true;
-    }
-
-    @Override
-    public boolean isOccupied(Vector2d position) {
-        for(Animal i : animals){
-            if(i.isAt(position))
-                return true;
-        }
-        return false;
+        return !(objectAt(dest) instanceof Animal) && LOWER_BOUND.precedes(dest) && upper_bound.follows(dest);
     }
 
     @Override
     public Object objectAt(Vector2d position) {
-        for(Animal i : animals){
-            if(i.isAt(position)){
-                return i;
-            }
-        }
-        return null;
+        return animals.stream()
```

Comment:
przy obecnym rozwiązaniu można by to przenieść do abstrakcyjnej klasie i potem w `Grass` korzystać z `super`